### PR TITLE
Add gatherv scatterv gtests

### PR DESCRIPTION
**Pull request by shimmybalsam**
## What
Adding gtests for gatherv and scatterv

### CODE REVIEW

**Comment by shimmybalsam** [comment]
bot:retest

File: test/gtest/coll/test_gatherv.cc [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+    const gtest_ucc_inplace_t  inplace  = std::get<4>(GetParam());
+    std::vector<UccReq>        reqs;
+    std::vector<UccCollCtxVec> ctxs;
+
+    for (int tid = 0; tid < UccJob::nStaticTeams; tid++) {
+        UccTeam_h     team = UccJob::getStaticTeams()[tid];
+        int           size = team->procs.size();
+        UccCollCtxVec ctx;
+
+        if (size == 1 && root > 0) {
```
Comment (samnordmann) [inline test/gtest/coll/test_gatherv.cc:262]: replace the predicate by
`size < root`
same in scatterv tests

This check should also be done in the other tests, right?

**Review by samnordmann — APPROVED** [review]

File: test/gtest/coll/test_scatterv.cc [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+            coll->flags             = 0;
+            coll->coll_type         = UCC_COLL_TYPE_SCATTERV;
+            coll->root              = root;
+            coll->dst.info.mem_type = mem_type;
+            coll->dst.info.count    = (ucc_count_t)my_count;
+            coll->dst.info.datatype = dtype;
+
+            if (r == root) {
+                all_counts = 0;
+                counts = (int*)malloc(sizeof(int) * nprocs);
```
[the remark this answers is not in the document]
Comment (shimmybalsam) [reply test/gtest/coll/test_scatterv.cc:44]: added EXPECT_NE

File: test/gtest/coll/test_gatherv.cc (hunk 1 above)
Comment (shimmybalsam) [reply test/gtest/coll/test_gatherv.cc]: Replaced with size <= root (also when size == root its invalid). Added in all gtests for gatherv and scatterv. Will add for all other collective gtests in seperate PR.

File: test/gtest/coll/test_gatherv.cc [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+            EXPECT_NE(ctxs[r]->init_buf, nullptr);
+            for (int i = 0; i < my_count * ucc_dt_size(dtype); i++) {
+                uint8_t *sbuf = (uint8_t *)ctxs[r]->init_buf;
+                sbuf[i]       = ((i + r) % 256);
+            }
+
+            if (r == root) {
+                all_counts = 0;
+                counts = (int*)malloc(sizeof(int) * nprocs);
+                displs = (int*)malloc(sizeof(int) * nprocs);
```
[the remark this answers is not in the document]
Comment (shimmybalsam) [reply test/gtest/coll/test_gatherv.cc:52]: added EXPECT_NE

File: test/gtest/coll/test_gatherv.cc [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+            coll->flags             = 0;
+            coll->coll_type         = UCC_COLL_TYPE_GATHERV;
+            coll->root              = root;
+            coll->src.info.mem_type = mem_type;
+            coll->src.info.count    = (ucc_count_t)my_count;
+            coll->src.info.datatype = dtype;
+
+            ctxs[r]->init_buf =
+                ucc_malloc(ucc_dt_size(dtype) * my_count, "init buf");
+            EXPECT_NE(ctxs[r]->init_buf, nullptr);
```
Comment (Sergei-Lebedev) [inline test/gtest/coll/test_gatherv.cc:42]: ASSERT_NE

File: test/gtest/coll/test_gatherv.cc [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+            for (int i = 0; i < my_count * ucc_dt_size(dtype); i++) {
+                uint8_t *sbuf = (uint8_t *)ctxs[r]->init_buf;
+                sbuf[i]       = ((i + r) % 256);
+            }
+
+            if (r == root) {
+                all_counts = 0;
+                counts = (int*)malloc(sizeof(int) * nprocs);
+                EXPECT_NE(counts, nullptr);
```
Comment (Sergei-Lebedev) [inline test/gtest/coll/test_gatherv.cc:51]: ASSERT_NE

File: test/gtest/coll/test_gatherv.cc [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+                displs = (int*)malloc(sizeof(int) * nprocs);
+                EXPECT_NE(displs, nullptr);
```
Comment (Sergei-Lebedev) [inline test/gtest/coll/test_gatherv.cc:53]: ASSERT_NE

File: test/gtest/coll/test_scatterv.cc [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+    }
+    void reset(UccCollCtxVec ctxs)
+    {
+        ucc_coll_args_t *coll;
+        ucc_datatype_t   dtype = (ctxs[0])->args->dst.info.datatype;
+        size_t           my_count;
+
+        for (auto r = 0; r < ctxs.size(); r++) {
+            coll = ctxs[r]->args;
+            my_count = coll->dst.info.count;
```
Comment (Sergei-Lebedev) [inline test/gtest/coll/test_scatterv.cc:121]: move my_count inside if() block

File: test/gtest/coll/test_scatterv.cc [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+        uint8_t *dsts;
+
+        for (auto r = 0; r < ctxs.size(); r++) {
+            if (r == root && inplace) {
+                continue;
+            }
+            my_count = (ctxs[r])->args->dst.info.count;
+            if (UCC_MEMORY_TYPE_HOST != mem_type) {
+                dsts = (uint8_t *)ucc_malloc(my_count * dt_size, "dsts buf");
+                EXPECT_NE(dsts, nullptr);
```
Comment (Sergei-Lebedev) [inline test/gtest/coll/test_scatterv.cc:144]: ASSERT_NE

File: test/gtest/coll/test_scatterv.cc [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+            }
+            if (UCC_MEMORY_TYPE_HOST != mem_type) {
```
Comment (Sergei-Lebedev) [inline test/gtest/coll/test_scatterv.cc:146]: can be combined with previous if

File: test/gtest/coll/test_scatterv.cc [hunk 10]
```diff
@@ -0,0 +1,309 @@
...
+                    my_count * ucc_dt_size(dtype),
+                    mem_type, 0);
+            }
+        }
+    }
+    bool data_validate(UccCollCtxVec ctxs)
+    {
+        bool     ret               = true;
+        int      root              = ctxs[0]->args->root;
+        size_t   dt_size = ucc_dt_size((ctxs[0])->args->dst.info.datatype);
```
Comment (Sergei-Lebedev) [inline test/gtest/coll/test_scatterv.cc:133]: alignment

File: test/gtest/coll/test_gatherv.cc [hunk 11]
```diff
@@ -0,0 +1,305 @@
...
+        ctxs.clear();
+    }
+    void reset(UccCollCtxVec ctxs)
+    {
+        ucc_coll_args_t *coll     = ctxs[root]->args;
+        size_t           my_count = coll->src.info.count;
+        ucc_datatype_t   dtype    = coll->dst.info_v.datatype;
+        int *            displs   = (int *)coll->dst.info_v.displacements;
+
+        clear_buffer(coll->dst.info_v.buffer, ctxs[root]->rbuf_size,
```
Comment (Sergei-Lebedev) [inline test/gtest/coll/test_gatherv.cc:124]: how clear_buffer works for non root ranks?

File: test/gtest/coll/test_scatterv.cc [hunk 12]
```diff
@@ -0,0 +1,309 @@
...
+                                        UCC_MEMORY_TYPE_HOST, mem_type));
+            } else {
+                dsts = (uint8_t *)ctxs[r]->args->dst.info.buffer;
+            }
+
+            for (int i = 0; i < my_count * dt_size; i++) {
+                if ((uint8_t)((i + r) % 256) !=
+                    dsts[i]) {
+                    ret = false;
+                    i   = my_count * dt_size;
```
Comment (Sergei-Lebedev) [inline test/gtest/coll/test_scatterv.cc:158]: break?

File: test/gtest/coll/test_gatherv.cc (hunk 11 above)
Comment (shimmybalsam) [reply test/gtest/coll/test_gatherv.cc:124]: coll is only for root, line 119: ucc_coll_args_t *coll     = ctxs[root]->args;

File: test/gtest/coll/test_scatterv.cc (hunk 8 above)
Comment (shimmybalsam) [reply test/gtest/coll/test_scatterv.cc]: changes to ucc_assert as ASSERT_NE can only be used inside void function and this data_validate function returns bool. Same for data_validate in test_gatherv.

**Review by Sergei-Lebedev — APPROVED** [review]

**Review by bureddy — APPROVED** [review]

**Comment by shimmybalsam** [comment]
bot:retest

**Pull request merged**